File: Longest Valid Parentheses/Longest Valid Parentheses.cpp
class Solution {
public:
    int longestValidParentheses(string s) {
        std::stack<int> stk;
        int start = 0;
        int len = 0;
        for(int i = 0; i < s.length(); ++i){
            char c = s[i];
            if(c == '('){
                stk.push(i);
            }
            else{
                if(stk.empty()){
                    // Start from next element
                    start = i + 1;
                }
                else{
                    stk.pop();
                    len = std::max(len, stk.empty() ? i - start + 1 : i - stk.top());
                }
            }
        }
        return len;
    }
};

File: Longest Valid Parentheses/Longest Valid Parentheses.java
public class Solution {
    public int longestValidParentheses(String s) {
        char[] c = s.toCharArray();
        
        int max = 0;
        int accu = 0;
        int numOfPairs = 0;
        for(int i = 0; i < c.length; ++i){
            switch(c[i]){
                case '(':
                    ++numOfPairs;
                    break;
                case ')':
                    if(numOfPairs == 0){
                        c[i] = '.';
                    }
                    else{
                        --numOfPairs;
                    }
                    break;
            }
        }
        
        numOfPairs = 0;
        for(int i = c.length - 1; i >= 0; --i){
            switch(c[i]){
                case ')':
                    ++numOfPairs;
                    ++accu;
                    break;
                case '(':
                    if(numOfPairs == 0){
                        c[i] = '.';
                        accu = 0;
                    }
                    else{
                        --numOfPairs;
                        ++accu;
                    }
                    break;
                case '.':
                    accu = 0;
                    break;
            }
            max = Math.max(max, accu);
        }
        
        return max;
    }
}
